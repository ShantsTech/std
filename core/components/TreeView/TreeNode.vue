<!--
/*
 * This file is part of Stants Tech LLC
 *
 * Copyright (C) 2020 Stants Tech LLC
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */
-->

<template>
  <li class="std-tree-node" :class="nodeClasses">
    <div
      :ref="cuid"
      v-if="showRoot"
      :class="nodeClasses"
      class="std-tree-node-wrapper"
    >
      <span v-if="hasChildren" class="std-tree-node-toggle">
        <std-icon-button
          @click="onClickNode"
          :name="isOpen ? 'chevron-up' : 'chevron-down'"
          role="none"
        />
      </span>
      <div class="std-tree-node-content">
        <slot name="content" :node-data="currentNode"></slot>
        <template v-if="!$slots['content']">
          {{ data.name }}
        </template>
      </div>
    </div>
    <div class="std-tree-node-liner" :style="lineStyles"></div>
    <transition
      :name="animation"
      v-if="hasChildren"
      @after-leave="onAnimationComplete"
    >
      <ul class="std-tree-node-child" v-if="isOpen">
        <std-tree-node
          v-for="(child, index) in data.children"
          :key="child.cuid"
          :data="child"
          :animation="animation"
          :isLast="index + 1 === data.children.length"
        >
          <template v-for="(_, name) in $slots" v-slot:[name]="slotData">
            <slot :name="name" v-bind="slotData" />
          </template>
        </std-tree-node>
      </ul>
    </transition>
  </li>
</template>

<script lang="ts">
import {defineComponent, PropType, ref} from 'vue';
import IconButton from '@sthrm/std/core/components/Button/Icon.vue';
import {TNode} from './types';
import {nanoid} from 'nanoid';

export default defineComponent({
  name: 'std-tree-node',
  props: {
    data: {
      type: Object as PropType<TNode>,
      required: true,
    },
    open: {
      type: Boolean,
      default: false,
    },
    showRoot: {
      type: Boolean,
      default: true,
    },
    animation: {
      type: String,
      default: 'transition-fade-down',
    },
    isLast: {
      type: Boolean,
      default: true,
    },
  },
  components: {
    'std-icon-button': IconButton,
  },
  setup(props) {
    const isOpen = ref(props.open);
    const isCollapsed = ref(!props.open);
    const height = ref(0);
    return {
      isOpen,
      isCollapsed,
      height,
    };
  },
  computed: {
    hasChildren(): boolean {
      return this.data.children && this.data.children.length;
    },
    nodeClasses(): object {
      return {
        '--parent': this.hasChildren,
        '--open': !this.isCollapsed && this.hasChildren,
        '--last': this.isLast,
      };
    },
    currentNode(): TNode {
      const {name, children, ...rest} = this.data;
      return {name, children, ...rest};
    },
    cuid(): string {
      return nanoid(8);
    },
    nodes(): TNode {
      return {...this.data, cuid: this.cuid};
    },
    lineStyles(): object {
      return {
        height: this.isLast ? this.height / 2 + 'px' : '100%',
      };
    },
  },
  methods: {
    onClickNode() {
      if (this.hasChildren) {
        this.isOpen = !this.isOpen;
        this.isCollapsed = false;
      }
    },
    onAnimationComplete() {
      this.isCollapsed = true;
    },
    calculateNodeHeight() {
      const node = this.$refs[this.cuid] as HTMLInputElement;
      this.height = node ? node.clientHeight : 0;
    },
  },
  mounted() {
    this.calculateNodeHeight();
  },
});
</script>

<style src="./tree-view.scss" lang="scss" scoped></style>
