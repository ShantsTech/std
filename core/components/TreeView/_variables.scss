// This file is part of Stants Tech LLC
//
// Copyright (C) 2020 Stants Tech LLC
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see  http://www.gnu.org/licenses

$std-tree-view-font-size: 12px !default;
$std-tree-view-font-color: $std-interface-gray-darken-1-color !default;
$std-tree-view-font-weight: 400 !default;
$std-tree-view-border-radius: 0.65rem !default;
$std-tree-view-line-height: 1.5 !default;
$std-tree-view-font-family: $std-font-family !default;

$std-tree-view-toggle-height: 1.2rem;
$std-tree-view-content-padding: 0.25rem 0.5rem;
$std-tree-view-node-padding: 0 0 0 1.5rem;
$std-tree-view-border: 0.15rem solid $std-primary-one-color;
$std-tree-view-animation-time: 500ms;
