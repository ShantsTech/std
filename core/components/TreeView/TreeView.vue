<!--
/*
 * This file is part of Stants Tech LLC
 *
 * Copyright (C) 2020 Stants Tech LLC
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */
-->

<template>
  <ul class="std-tree-view">
    <std-tree-node v-bind="$attrs">
      <template v-for="(_, name) in $slots" v-slot:[name]="slotData">
        <slot :name="name" v-bind="slotData" />
      </template>
    </std-tree-node>
  </ul>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import TreeNode from '@sthrm/std/core/components/TreeView/TreeNode.vue';

export default defineComponent({
  name: 'std-tree-view',
  inheritAttrs: false,
  components: {
    'std-tree-node': TreeNode,
  },
});
</script>

<style src="./tree-view.scss" lang="scss" scoped></style>
