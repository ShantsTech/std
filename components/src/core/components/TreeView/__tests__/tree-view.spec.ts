/*
 * This file is part of Shants Tech LLC
 *
 * Copyright (C) 2020 Shants Tech LLC
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */

import {mount} from '@vue/test-utils';
import TreeView from '@ohrm/std/core/components/TreeView/TreeView.vue';
import TreeNode from '@ohrm/std/core/components/TreeView/TreeNode.vue';
import {h} from 'vue';

const data = {
  name: 'base',
  children: [
    {name: 'root'},
    {
      name: 'usr',
      children: [
        {
          name: 'share',
          children: [{name: 'man'}, {name: 'local'}],
        },
        {name: 'bin'},
      ],
    },
    {name: 'etc'},
  ],
};

describe('TreeView.vue', () => {
  it('renders STD Tree View', () => {
    const wrapper = mount(TreeView, {
      props: {
        data: {
          name: 'test',
          children: [],
        },
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should renders Tree View with 4 Nodes', () => {
    const wrapper = mount(TreeView, {
      props: {data, open: true},
    });
    const nodes = wrapper.findAllComponents(TreeNode);
    expect(nodes.length).toBe(4);
  });

  it('should not render root node', () => {
    const wrapper = mount(TreeView, {
      props: {data, showRoot: false, open: true},
    });
    expect(wrapper.findAll('.std-tree-node-wrapper').length).toBe(3);
  });

  it('should open node on click', async () => {
    const wrapper = mount(TreeView, {
      props: {data},
    });
    wrapper.find('.std-icon-button').trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.std-tree-node').classes()).toContain('--open');
  });

  it('should not have collapse button on non parent nodes', async () => {
    const wrapper = mount(TreeView, {
      props: {data, open: true},
    });
    const nodes = wrapper.findAllComponents(TreeNode);
    expect(
      nodes[1]
        .find('.std-tree-node-wrapper')
        .find('.std-tree-node-toggle')
        .exists(),
    ).toBeFalsy();
  });

  it('should pass slot props to template', () => {
    const wrapper = mount(TreeView, {
      props: {data},
      slots: {
        content: ({nodeData}) => {
          return h('h1', nodeData.name);
        },
      },
    });
    const rootNode = wrapper.findComponent(TreeNode);
    expect(rootNode.find('.std-tree-node-content > h1').text()).toContain(
      'base',
    );
  });
});
