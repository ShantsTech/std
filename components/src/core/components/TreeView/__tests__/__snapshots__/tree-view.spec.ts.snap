// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeView.vue renders STD Tree View 1`] = `
<ul class="std-tree-view">
  <li class="std-tree-node --last">
    <div class="--last std-tree-node-wrapper">
      <!--v-if-->
      <div class="std-tree-node-content">test</div>
    </div>
    <div class="std-tree-node-liner" style="height: 0px;"></div>
    <!--v-if-->
  </li>
</ul>
`;
